src/meta/src/stream/stream_graph.rs

```diff
@@ -184,359 +175,399 @@ impl StreamActorBuilder {
-                        .upstreams
-                        .values()
-                        .filter(|x| x.same_worker_node)
-                        .exactly_one()
-                        .unwrap()
-                        .actors
-                        .as_global_ids()
-                        .into_iter()
-                        .exactly_one()
-                        .unwrap(),
+                assert!(!stream_node.get_fields().is_empty());
+                assert!(stream_node.input.is_empty());
+
+                // Index the upstreams by the an internal edge ID.
+                let upstreams = &self.upstreams[&EdgeId::Internal {
+                    link_id: stream_node.get_operator_id(),
+                }];
+
+                Ok(StreamNode {
+                    node_body: Some(NodeBody::Merge(MergeNode {
+                        upstream_actor_id: upstreams.actors.as_global_ids(),
+                        upstream_fragment_id: upstreams.fragment_id.as_global_id(),
+                        upstream_dispatcher_type: exchange.get_strategy()?.r#type,
+                        fields: stream_node.get_fields().clone(),
+                    })),
+                    identity: "MergeExecutor".to_string(),
+                    ..stream_node.clone()
                 })
-            } else {
-                None
-            },
-            vnode_bitmap: None,
-            // To be filled by `StreamGraphBuilder::build`
-            mview_definition: "".to_owned(),
+            }
+
+            // "Leaf" node `Chain`.
+            NodeBody::Chain(chain_node) => {
+                let input = stream_node.get_input();
+                assert_eq!(input.len(), 2);
+
+                let merge_node = &input[0];
+                assert_matches!(merge_node.node_body, Some(NodeBody::Merge(_)));
+                let batch_plan_node = &input[1];
+                assert_matches!(batch_plan_node.node_body, Some(NodeBody::BatchPlan(_)));
+
+                // Index the upstreams by the an external edge ID.
+                let upstreams = &self.upstreams[&EdgeId::UpstreamExternal {
+                    upstream_table_id: chain_node.table_id,
+                    downstream_fragment_id: self.fragment_id,
+                }];
+
+                // As we always use the `NoShuffle` exchange for MV on MV, there should be only one
+                // upstream.
+                let upstream_actor_id = upstreams.actors.as_global_ids();
+                assert_eq!(upstream_actor_id.len(), 1);
+
+                let chain_input = vec![
+                    // Fill the merge node with correct upstream info.
+                    StreamNode {
+                        input: vec![],
+                        stream_key: merge_node.stream_key.clone(),
+                        node_body: Some(NodeBody::Merge(MergeNode {
+                            upstream_actor_id,
+                            upstream_fragment_id: upstreams.fragment_id.as_global_id(),
```

Comment:
All fields for MV on MV can be filled directly in this phase, thanks to the info of external fragments in the "complete" graph.